### ISSUE TITLE
Troubles setting up on Windows

### ISSUE THREAD

**Issue by starry-abyss** [GodotNativeTools/godot-cpp#56]
I have VS2017 and Qt Creator as IDE. No matter how hard I try to launch `VsDevCmd.bat` before scons, the binding lib seems to be compiled as 64 bit (somehow it even compiles without pre-launching anything, unlike Godot).
When I compile the sample from readme, I get this first: 

![default](https://user-images.githubusercontent.com/11571820/33773546-551e87aa-dc49-11e7-9ae7-7756f39fae47.png)

Seems like `wchar_t` and `unsigned short` are the same on my platform.

If I comment out one of the operator declarations, I get a bunch of `unresolved external symbol` errors and the warning hint about 64 bits: 
![default](https://user-images.githubusercontent.com/11571820/33773643-a7bd676a-dc49-11e7-97cc-9f8072957093.png)

**Comment by karroffel** [comment]
Godot does a lot of things to detect the right MSVC version, this does nothing like that, and I honestly don't know how since I don't know much about Windows and also can't really test at the moment. Maybe someone else (@RichyHBM or @BastiaanOlij maybe?) who has more experience with Windows and MSVC can help here?

**Comment by BastiaanOlij** [comment]
I use 2017 without any problems. I just use the "x64 Native Tools Command Prompt" entry that VC adds. 
Note that this will indeed compile a 64bit version of the library and that is the only thing I've ever tested with the scons files we created. I'm not sure if any extra steps are needed to compile a 32bit version.

note that these lines of code are needed in the scons file:
env = Environment()
if platform == "windows":
    env = Environment(ENV = os.environ)

They were only recently added so possibly that may be related to your issue?

**Comment by starry-abyss** [comment]
@BastiaanOlij Thank you, yes, I do have those lines in the SConstruct file.

Here's diff between my env variables after "x64 Native Tools Command Prompt" and "x86 Native Tools Command Prompt": [patch.txt](https://github.com/GodotNativeTools/godot-cpp/files/1544585/patch.txt)

**BUT** the bindings still compile (64 bit) after `scons -c` from bare terminal without even launching any Visual Studio provided scripts!

**Comment by BastiaanOlij** [comment]
Hmmm, i'll have to do some experimenting tomorrow if I find time but note that I have never tried compiling the 32bit versions. Wouldn't be surprised if Microsoft has given up on those in 2017 or that it requires specific switches we never added.

**Comment by starry-abyss** [comment]
Can't make it compile with MinGW either, because of auto-detection of some VS. Going to try modules, maybe they are easier to start working with...

**Comment by dynamite-ready** [comment]
I'm experiencing further trouble on Windows 8:

Tried to build the GDNative CPP project. Following the instructions in the readme, I changed the SConstruct file with - `godot_bin_path = "c:/godot.bin"` - and then ran the scons from a Visual Studio command prompt, file with the following arguments - `scons godotbinpath="c:/godot/bin/" headers="../godot_headers/" p=windows generate_bindings=yes` - After doing so, I received this error message:

IOError: [Errno 2] No such file or directory: 'godot_api.json

It's strange though, as I think I managed to build the library recently (a 9mb .lib file, I think). Has something changed in either this project, the godot_headers project, or the main project?

**Comment by karroffel** [comment]
I think that error was resolved @dynamite-ready , right? Mind telling here how, in case other people run into that?

I'll close it anyway since I know from Discord that this has been fixed :)

**Comment by starry-abyss** [comment]
Did @dynamite-ready fix both his problem and the original problem of this thread?

**Comment by dynamite-ready** [comment]
My particular problem was down to not building the cpp-bindings project itself.
To fix that, I used the Sconstruct file in the project. 

If you're still having trouble with that, when running scons, for the 'binpath' value - `godot_bin_path = "..."` - Make sure this is the Godot .exe itself. Ideally the one in the bin folder found after building from source.

Also, to build the bindings, I used the 64bit VS command prompt (also make sure to run it as an admin). 

...

As for the issue in your OP ('unresolved external symbol * x'), I can't be entirely sure, as I'm not using VS, but it could most likely be down to not linking either the cpp-bindings .dll, the headers in the cpp-bindings/include folder, or godot_headers. 

So check the .dll first (it should be about 9mb on Windows if it compiles correctly), the paths to the 'cpp-bindings/include' that's created after compiling the bindings project, and the paths to the godot_header C files.

After that, I can't really help much with Visual Studio. I used Cmake to build my version of the example, and that abstracts away quite a lot of the configuration, compiler flags, etc.

**Comment by starry-abyss** [comment]
Glad that you got your issue fixed, and thank you for the tips!

My issue is about building the sample, not the bindings themselves (it's about bindings too, but only partially, see the part after quote below). Not sure if it's clear to @karroffel , because closing my issue just because yours is solved makes no sense to me.

> Also, to build the bindings, I used the 64bit VS command prompt
My issue is that it always builds in 64 bit even without any VS command prompts...

That said, I'm going to use modules instead, they use only one build system, which is at the moment far more convenient than having 3 build systems (engine+bindings+actual project), which in the end happen to be not compatible with each other, because some VS flags/whatever :-)

TL;DR I don't mind that the issue is closed, just remember it wasn't solved

**Comment by karroffel** [comment]
Ups, I didn't notice it wasn't your issue. 😅

**Comment by dynamite-ready** [comment]
> My issue is about building the sample, not the bindings themselves (it's about bindings too, but only partially, see the part after quote below).

Same here. I thought I managed to compile the binding the first time, but what actually compiled was a dead .dll, and that was at the root of my problem in getting the sample to work.

This specific issue about compiling a 32bit build though... Change the title, perhaps?

**Comment by starry-abyss** [comment]
Yes, I'm using 32-bit Godot. At the moment of asking, it was the only variant I compiled, and the interface of passing path to Godot was a bit different IIRC.

> Change the title, perhaps?

IDK.

**Comment by Zylann** [comment]
> did you link to the 32bit version of Godot in your bin path when compiling the bindings?

@dynamite-ready you should not need to link to Godot though

**Comment by starry-abyss** [comment]
I'm sure, dynamite-ready meant "pass the path to Godot executable" when he said "link".

**Comment by BastiaanOlij** [comment]
I just ran through the steps as explained on the current readme.md in master and that worked like a charm. Do note the new `bits=32` / `bits=64` switch for scons, it attempts to auto detect but its not perfect (yet) especially when using MSVC so use the switches when not sure.

Closing this as there has been no discussion for nearly a year.

**Comment by starry-abyss** [comment]
Thank you! I'm a happy modules user now, but surely this would help other people struggling.
